File: cloudflare-relay/src/apns.ts
/**
 * APNs background-push helper — sends a single content-available
 * silent push to one device token via Apple's HTTP/2 API. Used by
 * sync durability Phase 3c to wake the iOS app when new ops land
 * in the relay so it can pull before the OS suspends it.
 *
 * Cloudflare Workers runtime only: Web Crypto (`crypto.subtle`),
 * `fetch`, `TextEncoder`, `atob`/`btoa`. No Node APIs.
 */

export interface ApnsEnv {
  APNS_KEY_P8?: string;    // PEM contents of the APNs auth key (.p8)
  APNS_KEY_ID?: string;    // 10-char APNs key id (JWT header `kid`)
  APNS_TEAM_ID?: string;   // 10-char Apple team id (JWT claim `iss`)
  APNS_BUNDLE_ID?: string; // e.g. "app.tesela.ios" — the `apns-topic`
  APNS_HOST?: string;      // optional override; default "https://api.push.apple.com"
}

/** Module-level JWT cache — APNs allows token reuse up to ~60 min;
 *  we re-mint after 50 min to stay well inside the window. Keyed by
 *  kid so swapping keys (rotation) invalidates automatically. */
let jwtCache: { kid: string; token: string; expiresAt: number } | null = null;
const JWT_CACHE_TTL_MS = 50 * 60 * 1000;

/**
 * Sends ONE content-available background push to `deviceTokenHex`.
 * Returns true on a 2xx from APNs; false on ANY failure OR if APNs
 * is not configured (any required env field missing). MUST NOT throw.
 */
export async function sendApnsBackgroundPush(
  env: ApnsEnv,
  deviceTokenHex: string,
): Promise<boolean> {
  try {
    // Config guard — the relay runs fine before the key is provisioned.
    const { APNS_KEY_P8, APNS_KEY_ID, APNS_TEAM_ID, APNS_BUNDLE_ID } = env;
    if (!APNS_KEY_P8 || !APNS_KEY_ID || !APNS_TEAM_ID || !APNS_BUNDLE_ID) {
      return false;
    }

    const token = await getOrMintJwt(APNS_KEY_P8, APNS_KEY_ID, APNS_TEAM_ID);
    const host = env.APNS_HOST || "https://api.push.apple.com";

    const res = await fetch(`${host}/3/device/${deviceTokenHex}`, {
      method: "POST",
      headers: {
        authorization: `bearer ${token}`,
        "apns-topic": APNS_BUNDLE_ID,
        "apns-push-type": "background",
        "apns-priority": "5",
        "apns-expiration": "0",
        "content-type": "application/json",
      },
      body: JSON.stringify({ aps: { "content-available": 1 } }),
    });
    return res.ok;
  } catch {
    // ANY unexpected error (crypto, network, parse) → false, never throw.
    return false;
  }
}

// ─── JWT (ES256 provider token) ───────────────────────────────────

async function getOrMintJwt(
  keyP8: string,
  kid: string,
  teamId: string,
): Promise<string> {
  const now = Date.now();
  if (jwtCache && jwtCache.kid === kid && now < jwtCache.expiresAt) {
    return jwtCache.token;
  }
  const token = await mintJwt(keyP8, kid, teamId);
  jwtCache = { kid, token, expiresAt: now + JWT_CACHE_TTL_MS };
  return token;
}

async function mintJwt(
  keyP8: string,
  kid: string,
  teamId: string,
): Promise<string> {
  const header = base64urlEncode(
    new TextEncoder().encode(JSON.stringify({ alg: "ES256", kid })),
  );
  const iat = Math.floor(Date.now() / 1000);
  const payload = base64urlEncode(
    new TextEncoder().encode(JSON.stringify({ iss: teamId, iat })),
  );
  const signingInput = `${header}.${payload}`;

  const der = pemToDer(keyP8);
  const cryptoKey = await crypto.subtle.importKey(
    "pkcs8",
    der,
    { name: "ECDSA", namedCurve: "P-256" },
    false,
    ["sign"],
  );
  // Web Crypto ECDSA returns raw r||s (IEEE P1361, 64 bytes for P-256)
  // — that IS the JWT ES256 signature. Base64url-encode directly.
  const sigBuf = await crypto.subtle.sign(
    { name: "ECDSA", hash: "SHA-256" },
    cryptoKey,
    new TextEncoder().encode(signingInput),
  );
  const signature = base64urlEncode(new Uint8Array(sigBuf));
  return `${signingInput}.${signature}`;
}

// ─── Encoding helpers ──────────────────────────────────────────────

/** Base64url (RFC 7515 §2): no `+`, `/`, or `=` padding. */
function base64urlEncode(bytes: Uint8Array): string {
  let bin = "";
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]!);
  return btoa(bin)
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

/** Strip PEM armor + whitespace, then base64-decode to raw DER bytes. */
function pemToDer(pem: string): Uint8Array {
  const b64 = pem
    .replace(/-----BEGIN PRIVATE KEY-----/, "")
    .replace(/-----END PRIVATE KEY-----/, "")
    .replace(/\s/g, "");
  const bin = atob(b64);
  const der = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) der[i] = bin.charCodeAt(i);
  return der;
}

File: crates/tesela-relay/src/apns.rs
//! APNs silent-push proxy (sync durability P3c).
//!
//! Sends a single content-available ("background") push to one device
//! token so a suspended iOS device wakes and drains pending ops from the
//! relay. This is the Rust port of the Cloudflare Worker's `apns.ts` —
//! the SAME JWT structure, request headers, and body, so either relay
//! drives the same iOS receiver (the `register_device` POST that iOS
//! build 36 already ships authenticates here unchanged).
//!
//! Zero-knowledge: the push carries NO note content (only
//! `content-available: 1`) — it just nudges the device to pull the
//! already-AEAD-sealed ops it would have polled for anyway.
//!
//! Best-effort: a single failed push is not a sync failure (the device
//! still polls / runs its BGProcessingTask), so nothing here returns an
//! error into the deposit path — it logs at routing level and returns a
//! bool. The whole feature is optional: with any `APNS_*` value unset,
//! `from_config` returns `None` and the relay runs exactly as before.

use std::sync::Mutex;
use std::time::Duration;

use jsonwebtoken::{encode, Algorithm, EncodingKey, Header};
use serde::Serialize;

const DEFAULT_HOST: &str = "https://api.push.apple.com";
/// Apple allows provider-token reuse up to ~60 min and rejects more than
/// one mint per 20 min (`TooManyProviderTokenUpdates`), so cache ~50 min.
const JWT_TTL_SECS: i64 = 50 * 60;

/// A configured APNs sender. Built ONCE at startup (a bad `.p8` fails
/// there, never per-push) and shared via `AppState`. Holds a pooled
/// HTTP/2 client to APNs + the parsed ES256 signing key + a cached
/// provider token.
pub struct Apns {
    http: reqwest::Client,
    encoding_key: EncodingKey,
    key_id: String,
    team_id: String,
    bundle_id: String,
    host: String,
    /// Cached provider-token JWT + its expiry (unix secs).
    jwt_cache: Mutex<Option<(String, i64)>>,
}

#[derive(Serialize)]
struct Claims {
    iss: String,
    iat: i64,
}

impl Apns {
    /// Build an ENABLED sender only if all four `APNS_*` values are
    /// present AND the `.p8` parses; otherwise `None` (the relay then
    /// runs with no push, exactly as before — never an error).
    ///
    /// `key_p8` is auto-detected: inline PEM if it starts with
    /// `-----BEGIN`, otherwise a filesystem path to the `.p8` (the
    /// HA-add-on-friendly form — mount the file, point the env at it).
    /// Never panics; never logs the key bytes.
    pub fn from_config(
        key_p8: Option<String>,
        key_id: Option<String>,
        team_id: Option<String>,
        bundle_id: Option<String>,
        host: Option<String>,
    ) -> Option<Self> {
        let nonempty = |s: Option<String>| s.filter(|v| !v.trim().is_empty());
        let (key_p8, key_id, team_id, bundle_id) = match (
            nonempty(key_p8),
            nonempty(key_id),
            nonempty(team_id),
            nonempty(bundle_id),
        ) {
            (Some(a), Some(b), Some(c), Some(d)) => (a, b, c, d),
            _ => return None,
        };

        let pem: Vec<u8> = if key_p8.trim_start().starts_with("-----BEGIN") {
            key_p8.into_bytes()
        } else {
            match std::fs::read(&key_p8) {
                Ok(bytes) => bytes,
                Err(e) => {
                    tracing::warn!("APNs: cannot read key file {key_p8}: {e}; push disabled");
                    return None;
                }
            }
        };
        let encoding_key = match EncodingKey::from_ec_pem(&pem) {
            Ok(k) => k,
            Err(e) => {
                tracing::warn!("APNs: invalid .p8 EC key: {e}; push disabled");
                return None;
            }
        };
        let http = match reqwest::Client::builder()
            // Bound a hung APNs POST so a stalled push can never wedge the
            // background task that drives it. HTTP/2 (the only protocol
            // APNs speaks) is negotiated via ALPN by the http2 feature.
            .timeout(Duration::from_secs(10))
            .build()
        {
            Ok(c) => c,
            Err(e) => {
                tracing::warn!("APNs: HTTP client build failed: {e}; push disabled");
                return None;
            }
        };

        Some(Self {
            http,
            encoding_key,
            key_id,
            team_id,
            bundle_id,
            host: nonempty(host).unwrap_or_else(|| DEFAULT_HOST.to_string()),
            jwt_cache: Mutex::new(None),
        })
    }

    /// Mint-or-reuse the ES256 provider token. `jsonwebtoken` emits the
    /// JWS with the ES256 signature already in JOSE raw-`r||s` form — the
    /// same structure the Worker hand-rolls via WebCrypto.
    fn provider_token(&self) -> Result<String, jsonwebtoken::errors::Error> {
        let now = chrono::Utc::now().timestamp();
        {
            let cache = self.jwt_cache.lock().expect("apns jwt cache poisoned");
            if let Some((tok, exp)) = cache.as_ref() {
                if now < *exp {
                    return Ok(tok.clone());
                }
            }
        }
        let mut header = Header::new(Algorithm::ES256);
        // Drop the default `typ: "JWT"` so the header is byte-identical to
        // the CF Worker's {alg, kid} (Apple tolerates typ, but parity keeps
        // the two relays interchangeable).
        header.typ = None;
        header.kid = Some(self.key_id.clone());
        let claims = Claims {
            iss: self.team_id.clone(),
            iat: now,
        };
        let token = encode(&header, &claims, &self.encoding_key)?;
        *self.jwt_cache.lock().expect("apns jwt cache poisoned") =
            Some((token.clone(), now + JWT_TTL_SECS));
        Ok(token)
    }

    /// Send ONE content-available push to `device_token_hex`. Returns
    /// `true` on a 2xx from APNs, `false` on any failure (config, crypto,
    /// network, non-2xx). Never panics. Logs at routing level only — a
    /// token PREFIX, the status, and Apple's `reason` code — never the
    /// full token, the JWT, or the key; the push body has no note content.
    pub async fn send_background_push(&self, device_token_hex: &str) -> bool {
        let tag = &device_token_hex[..device_token_hex.len().min(8)];
        let token = match self.provider_token() {
            Ok(t) => t,
            Err(e) => {
                tracing::error!("[apns] push {tag}… jwt error: {e}");
                return false;
            }
        };
        let url = format!("{}/3/device/{}", self.host, device_token_hex);
        let res = self
            .http
            .post(&url)
            .header("authorization", format!("bearer {token}"))
            .header("apns-topic", &self.bundle_id)
            .header("apns-push-type", "background")
            .header("apns-priority", "5")
            .header("apns-expiration", "0")
            .header("content-type", "application/json")
            .body(r#"{"aps":{"content-available":1}}"#)
            .send()
            .await;
        match res {
            Ok(r) if r.status().is_success() => {
                tracing::info!("[apns] push {tag}… → {} OK", r.status().as_u16());
                true
            }
            Ok(r) => {
                let status = r.status().as_u16();
                // APNs returns {"reason":"BadDeviceToken"|...} on failure —
                // an error CODE, never note content. Surfacing it is what
                // makes the relay logs diagnostic.
                let reason = r
                    .text()
                    .await
                    .ok()
                    .and_then(|b| serde_json::from_str::<serde_json::Value>(&b).ok())
                    .and_then(|v| {
                        v.get("reason")
                            .and_then(|x| x.as_str().map(str::to_string))
                    })
                    .unwrap_or_default();
                tracing::warn!(
                    "[apns] push {tag}… → {status} FAIL reason={}",
                    if reason.is_empty() {
                        "?".to_string()
                    } else {
                        reason
                    }
                );
                false
            }
            Err(e) => {
                tracing::error!("[apns] push {tag}… error: {e}");
                false
            }
        }
    }
}
